instrumentation/elasticsearch/elasticsearch-rest-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/elasticsearch/rest/ElasticsearchClientAttributeExtractor.java

```diff
@@ -0,0 +1,124 @@
+    internalSet(attributes, SemanticAttributes.HTTP_METHOD, request.getMethod());
+    setPathPartsAttributes(attributes, request);
+  }
+
+  @Override
+  public void onEnd(
+      AttributesBuilder attributes,
+      Context context,
+      ElasticsearchRestRequest request,
+      @Nullable Response response,
+      @Nullable Throwable error) {
+    if (response != null) {
+      HttpHost host = response.getHost();
+      String uri = response.getRequestLine().getUri();
+      uri = uri.startsWith("/") ? uri : "/" + uri;
+      String url = host.toURI() + uri;
+
+      internalSet(attributes, SemanticAttributes.HTTP_URL, url);
+
+      if (host.getAddress() != null) {
+        internalSet(
+            attributes, SemanticAttributes.NET_PEER_NAME, host.getAddress().getHostAddress());
+        internalSet(attributes, SemanticAttributes.NET_PEER_PORT, (long) host.getPort());
+        if (host.getAddress() instanceof Inet6Address) {
+          internalSet(attributes, SemanticAttributes.NET_SOCK_FAMILY, "inet6");
+        }
+      }
+    }
+  }
+
+  private void setPathPartsAttributes(
+      AttributesBuilder attributes, ElasticsearchRestRequest request) {
+    ElasticsearchEndpointDefinition endpointDef = request.getEndpointDefinition();
+    if (endpointDef == null) {
+      return;
+    }
+    String[] availableRoutes = endpointDef.getRoutes();
+    if (availableRoutes == null || availableRoutes.length == 0) {
+      return;
+    }
+    String urlPath = request.getEndpoint();
+    boolean routeFound = false;
+    for (int i = 0; i < availableRoutes.length && !routeFound; i++) {
+      String route = availableRoutes[i];
+      if (route.contains("{")) {
+        Matcher matcher = matchUrl(route, urlPath);
+        if (matcher.find()) {
+          List<String> pathPartNames = routePathPartNames.get(route);
+          if (pathPartNames != null) {
+            for (String key : pathPartNames) {
+              try {
+                String value = matcher.group(key);
+                String attributeKey = "db.elasticsearch.path_parts." + key;
+                internalSet(attributes, AttributeKey.stringKey(attributeKey), value);
+              } catch (RuntimeException e) {
+                // ignore
+              }
+            }
+          }
+
+          routeFound = true;
+        }
+      }
+    }
+  }
+
+  private Matcher matchUrl(String route, String urlPath) {
+    if (!regexPatternMap.containsKey(route)) {
+
+      String regexStr = '^' + route.replace("{", "(?<").replace("}", ">[^/]+)") + '$';
+      regexPatternMap.put(route, Pattern.compile(regexStr));
+
+      if (route.contains("{")) {
+        List<String> pathPartNames = new ArrayList<>();
+        Matcher matcher = pathPartNamesPattern.matcher(route);
+        while (matcher.find()) {
+          try {
+            pathPartNames.add(matcher.group(1));
+          } catch (RuntimeException e) {
+            // ignore
```

Comment:
as you have a fixed set of routes does an exception hear mean that one of the routes is "malformed"? If so couldn't you verify in unit test that all the routes area good?

Reply:
There are more than 400 routes, not sure we want to test all of them explicitly?

Added this try - catch to prevent the instrumentation from failing completely in case a single path part cannot be retrieved.

Reply:
If you believe that there could be an exception then testing all the 400 routes seems reasonable. 400 is a small enough number so that there shouldn't be no issues testing them all. Just extract the part that deals with this into a package private method and create a unit test for it that calls this method with all possible routes. Unit testing javaagent modules is a bit of a pain because you need to have the test in a separate module so it wouldn't run with the agent. See https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/couchbase/couchbase-2-common/javaagent-unit-tests for inspiration

Reply:
added Unit test for testing all of the endpoints